lib/readPackage.nix

```diff
@@ -21,61 +21,47 @@
   ...
 } @ buildOptions: drv: let
   inherit (self.lib) inspectBuild;
-
-  element = rec {
-    active = true;
-    inherit attrPath;
-    # TODO deduplicate with logic for floxEnvs
-    # normalize to include "flake:", which is included in manifest.json
-    originalUrl =
-      if flakeRef == "self"
-      then "." # TODO: use outPath?
-      else
-        (
-          if flakeRef == null || builtins.match ".*:.*" flakeRef == []
-          then flakeRef
-          else "flake:${flakeRef}"
-        );
-    # TODO deduplicate with logic for floxEnvs and figure out a better way to
-    # store flake resolution information
-    url =
-      if flakeRef == "self"
-      then ""
-      else if flakeRef != null
-      then let
-        flake =
-          builtins.getFlake flakeRef;
-        # this assumes that either flakeRef is not indirect, or if it is indirect, the flake it
-        # resolves to contains a branch
-      in "${originalUrl}/${flake.rev}"
-      # TODO this violates the catalog schema, so it must be set with
-      # postprocessing
-      else null;
-    storePaths =
-      if drv.meta ? outputsToInstall
-      then
-        # only include outputsToInstall
-        (builtins.map (outputName: eval.outputs.${outputName})
-          drv.meta.outputsToInstall)
-      else lib.attrValues eval.outputs;
-  };
+  lockedFlake = (import ./lockFlake.nix).lockFlake flakeRef;
```

Comment:
I'm not quite sure I understand what you mean. Could you recommend a snippet?

Reply:
Thinking of something like this:

```suggestion
inherit (rec {
  lockedFlake = (import ./lockFlake.nix).lockFlake flakeRef;
  originalUrl = if flakeRef == null   then null else
                  if flakeRef == "self" then "."  else
                  lockedFlake.originalRef.string;
  url = if flakeRef == null   then null else
          if flakeRef == "self" then ""   else
          lockedFlake.lockedRef.string;
}) originalUrl url;
```

Its just preventing you from using `lockedFlake` without checking for the `null` case and shooting yourself in the foot in other places.

Maybe I've been doing too much rust where something like

```rust
let a = { 
	let f = File::open(...);
    let mut buf = String::new();
	f.read_all(&mut buf);
    buf
};

is a more common pattern.

Ignore this if you think its overkill

Reply:
Nix is lazy, deeply nested if blocks aren't necessary.
Nested `let ... in` expressions are only a relevant optimization in incredibly large files.

Not that it really matters but `fn1` allocates quite a bit more than `fn2` in order to form a new `nix::Bindings` value instead of just adding an extra `nix::Attr` to an existing `Bindings` set.
```
fn1 = i: let
  x = let a = 1; b = 2; in a + b;
  z = 3;
  y = z + 4;
in if i == 2 then x else y;

fn2 = i: let
  a = 1;
  b = 2;
  x = a + b;
  z = 3;
  y = z + 4;
in if i == 2 then x else y;
```

In any case, don't over optimize, the evaluator is lazy and nested `if/let` blocks just make your code harder to read.